sycl-fusion/passes/target/TargetFusionInfo.cpp

```diff
@@ -281,6 +485,243 @@ class NVPTXTargetFusionInfo : public TargetFusionInfoImpl {
+    if (Name.consume_front("ntid."))
+      return BuiltinKind::LocalSizeRemapper;
+    if (Name.consume_front("nctaid."))
+      return BuiltinKind::NumWorkGroupsRemapper;
+    return {};
+  }
+
+  bool shouldRemap(BuiltinKind K, const NDRange &SrcNDRange,
+                   const NDRange &FusedNDRange) const override {
+    switch (K) {
+    case BuiltinKind::LocalSizeRemapper:
+    case BuiltinKind::NumWorkGroupsRemapper:
+    case BuiltinKind::GlobalOffsetRemapper:
+      return true;
+    case BuiltinKind::LocalIDRemapper:
+    case BuiltinKind::GroupIDRemapper:
+      // If the local size is not specified, we have to unconditionally remap
+      // local and group IDs to guarantee correct global IDs.
+      return !SrcNDRange.hasSpecificLocalSize() ||
+             requireIDRemapping(SrcNDRange, FusedNDRange);
+    default:
+      llvm_unreachable("Unhandled kind");
+    }
+  }
+
+  bool isSafeToNotRemapBuiltin(Function *F) const override {
+    // `SubgroupLocalInvocationId` lowers to the `laneid`.
+    // Other subgroup-related builtins are computed from standard getters
+    // (workgroup size, local ID etc.) and constants (subgroup max size := 32),
+    // so we can't filter them out here.
+    return F->getIntrinsicID() != Intrinsic::nvvm_read_ptx_sreg_laneid;
+  }
+
+  unsigned getIndexSpaceBuiltinBitwidth() const override { return 32; }
+
+  void setMetadataForGeneratedFunction(Function *F) const override {
+    auto &Ctx = F->getContext();
+    if (F->getName().contains("__global_offset_remapper")) {
+      F->setAttributes(AttributeList::get(
+          Ctx,
+          AttributeSet::get(
+              Ctx, {Attribute::get(Ctx, Attribute::AttrKind::NoUnwind),
+                    Attribute::get(Ctx, Attribute::AttrKind::Speculatable),
+
+                    Attribute::get(Ctx, Attribute::AttrKind::AlwaysInline)}),
+          {}, {}));
+    } else {
+      F->setAttributes(AttributeList::get(
+          Ctx,
+          AttributeSet::get(
+              Ctx, {Attribute::get(Ctx, Attribute::AttrKind::NoCallback),
+                    Attribute::get(Ctx, Attribute::AttrKind::NoFree),
+                    Attribute::get(Ctx, Attribute::AttrKind::NoSync),
+                    Attribute::get(Ctx, Attribute::AttrKind::NoUnwind),
+                    Attribute::get(Ctx, Attribute::AttrKind::Speculatable),
+                    Attribute::get(Ctx, Attribute::AttrKind::WillReturn),
+                    Attribute::get(Ctx, Attribute::AttrKind::AlwaysInline)}),
+          {}, {}));
+    }
+    F->setMemoryEffects(MemoryEffects::none());
+  }
+
+  Value *getGlobalIDWithoutOffset(IRBuilderBase &Builder,
+                                  const NDRange &FusedNDRange,
+                                  uint32_t Idx) const override {
+    // Construct (or reuse) a helper function to query the global ID.
+    std::string GetGlobalIDName =
+        ("__global_id_" + Twine(static_cast<char>('x' + Idx))).str();
```

Comment:
```suggestion
    std::string GetGlobalIDName =
        Twine("__global_id_").concat(static_cast<char>('x' + Idx)).str();
```
I think this makes better use of `Twine`

Reply:
Had to revert, `concat` only accepts a `Twine`, so it would have to be `.concat(Twine('x'))`, which I find less readable than just `+ Twine('x')`.

Reply:
Fine with LGTM